pkg/build/builder/docker.go

```diff
@@ -80,23 +83,53 @@ func (d *DockerBuilder) Build() error {
 		push = true
 	}
 
-	if err := d.dockerBuild(buildDir, d.build.Spec.Source.Secrets); err != nil {
+	// TODO remove the code duplication in docker.go and sti.go.
+
+	// TODO replace this with a truly random string
+	suffix := fmt.Sprintf("%x", time.Now().UnixNano())
+	// buildTag is a random tag used for building the image in such a way
+	// that we can refer to the built image knowing that it was built here
+	// and not concurrently by a different goroutine or process.
+	buildTag := namer.GetName(pushTag, suffix, reference.NameTotalLengthMax)
+
+	if err := d.dockerBuild(buildDir, buildTag, d.build.Spec.Source.Secrets); err != nil {
+		return err
+	}
+
+	// TODO Shall we remove this tag/image or just leave it there? If we
+	// don't get to tag with the pushTag below, the image will be gone.
+	defer removeImage(d.dockerClient, buildTag)
```

Comment:
I'm normally writing "random" with quotes, because in fact it is the same tag as the original, plus a random suffix, as random as pod names are.

Given an original repo/tag "172.30.124.47:5000/sample-app/ruby-hello-world:latest", a build tag will look like "172.30.124.47:5000/sample-app/ruby-hello-world:latest-9acb0442".